### PR TITLE
[RFR] Code splitting and ES6

### PR DESCRIPTION
**Pull request by jpetitcolas**
- [x] Split code into several smaller fragments
- [x] Fix chart display and zoom events
- [x] Split `Drawer` into several sub-classes
- [x] Inverse Zoomer/ZoomArea relationship
- [x] Fix tests

### CODE REVIEW

File: lib/xAxis.js [hunk 1]
```diff
@@ -1,44 +1,17 @@
...
-    .tickFormat(tickFormat)
-  ;
-
-  if (typeof config.axisFormat === 'function') {
-    config.axisFormat(xAxis);
-  }
-
-  var y = (where == 'bottom' ? parseInt(graphHeight) : 0) + config.margin.top - 40;
-
-  xAxisEls[where] = graph
```
Comment (fzaninotto) [inline lib/xAxis.js]: This seems to have disappeared in the refactoring

Comment (jpetitcolas) [inline lib/xAxis.js]: No, it is now in the `Drawer`. An `xAxis` should not keep track of a collection of axes. 

By the way, the `Drawer` is far too heavy. It will be splitted in several smaller responsabilities classes.

**Comment by jpetitcolas** [comment]
Tests are green, and ready for review!

File: test/karma/eventDrops.js [hunk 2]
```diff
@@ -1,100 +1,119 @@
...
-      expect(lines.length).toBe(1);
+        expect(div.querySelectorAll('.drop').length).toBe(3);
     });
 
-  });
-
-  it ('should have as many circle by name as timestamps', function () {
-    elements.each(function (data) {
-      var lines = d3.select(this).select('svg').selectAll('.line');
+    /**
```
Comment (jpetitcolas) [inline test/karma/eventDrops.js]: Commented out code

File: lib/drawer/line.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
+import { drops } from './drops';
+import { labels } from './labels';
+
+export const line = function (svg, xScale, yScale, data, eventColor = '#000') {
```
Comment (djhi) [inline lib/drawer/line.js]: why not an arrow function here too ?

File: lib/drawer/xAxis.js [hunk 4]
```diff
@@ -0,0 +1,12 @@
+import xAxis from '../xAxis';
+
+const drawAxis = function (svg, xScale, configuration, orientation, y) {
```
Comment (djhi) [inline lib/drawer/xAxis.js]: why not an arrow function here too ?

File: lib/filterData.js [hunk 5]
```diff
@@ -1,19 +1,6 @@
...
-      return;
-    }
-    filteredData.push(datum);
-  });
-
-  return filteredData;
-};
+export default function filterData(data = [], scale) {
+    const [min, max] = scale.domain();
+    return data.filter(d => {
```
Comment (djhi) [inline lib/filterData.js]: on one line ?

**Comment by djhi** [comment]
:+1: es6 ftw ! Great job :)

**Comment by jpetitcolas** [comment]
Code review applied.

File: demo/demo.js [hunk 6]
```diff
@@ -1,3 +1,5 @@
+/*eslint-disable */
+
 // create dataset
 var data = [];
```
Comment (brikou) [inline demo/demo.js]: es3?

File: lib/configure.js [hunk 7]
```diff
@@ -0,0 +1,13 @@
+export default (config, defaultConfig) => {
+    const finalConfiguration = {};
```
Comment (brikou) [inline lib/configure.js]: Do you mean?

```js
export default (config, defaultConfig) => ({...defaultConfig, ...config});
```

If so, is this file really needed?

File: lib/drawer/drops.js [hunk 8]
```diff
@@ -0,0 +1,29 @@
+import filterData from '../filterData';
+
+export default (svg, scales, configuration) => function (data) {
```
Comment (brikou) [inline lib/drawer/drops.js]: IMVHO, using explicit configuration args would make it more readable, see...

```js
export default (svg, scales, { configuration: { eventColor, eventLineColor, lineHeight } }) => data => {
```

File: lib/drawer/index.js [hunk 9]
```diff
@@ -0,0 +1,50 @@
+import d3 from 'd3';
+import { delimiters } from './delimiters';
+import filterData from '../filterData';
+import { init as initMetaballs } from '../metaballs';
+
+export default function (svg, dimensions, scales, configuration) {
```
Comment (brikou) [inline lib/drawer/index.js]: Please exand configuration, as signature would be really meaningful, please, please

File: lib/drawer/index.js [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        .attr('width', dimensions.width)
+        .attr('height', 30)
+        .attr('transform', `translate(${configuration.margin.left}, ${configuration.margin.top - 45})`);
+
+    configuration.metaballs && initMetaballs(defs);
+
+    const lineSeparator = require('./lineSeparator')(axesContainer, scales, configuration, dimensions);
+    const axes = require('./axes')(axesContainer, scales, configuration, dimensions);
+    const labels = require('./labels')(labelsContainer, scales, configuration);
+    const drops = require('./drops')(dropsContainer, scales, configuration);
```
Comment (brikou) [inline lib/drawer/index.js]: Please expand configuration arg.

File: package.json [hunk 11]
```diff
@@ -8,18 +8,27 @@
     "url": "https://github.com/marmelab/EventDrops"
   },
   "dependencies": {
-    "d3": "3.5.3"
+    "babel-loader": "^5.4.0",
```
Comment (Kmaschta) [inline package.json]: You should `~` instead of all `^`.

File: lib/drawer/index.js [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+    return function (data) {
```
Comment (brikou) [inline lib/drawer/index.js]: FAT arrows are nice.

File: lib/drawer/line.js [hunk 13]
```diff
@@ -0,0 +1,14 @@
+import { drops } from './drops';
+import { labels } from './labels';
+
+export const line = (svg, xScale, yScale, data, eventColor = '#000') => {
+    var dropElements = svg.selectAll('g').data(data);
```
Comment (brikou) [inline lib/drawer/line.js]: var is bad

File: lib/metaballs.js [hunk 14]
```diff
@@ -0,0 +1,22 @@
+export const init = defs => {
```
Comment (brikou) [inline lib/metaballs.js]: `s/init/metaballs/`

File: lib/util/configurable.js [hunk 15]
```diff
@@ -1,16 +1,17 @@
...
+      (item => {
+          targetFunction[item] = value => {
+              if (!arguments.length) return config[item];
 
-        return targetFunction;
-      };
-    })(item); // for doesn't create a closure, forcing it
+              config[item] = value;
+              if (listeners.hasOwnProperty(item)) {
+                listeners[item](value);
```
Comment (brikou) [inline lib/util/configurable.js]: indent

File: package.json (hunk 11 above)
Comment (RobinBressan) [inline package.json]: Or none of them to prevent future issues

File: demo/demo.js (hunk 6 above)
Comment (jpetitcolas) [inline demo/demo.js]: I kept voluntarily the demo in ES5 to widen our target. It is then ready to copy/paste without having to configure any transpilers.

File: lib/configure.js (hunk 7 above)
Comment (jpetitcolas) [inline lib/configure.js]: Indeed. File going to be removed in next commit.

File: lib/drawer/drops.js (hunk 8 above)
Comment (jpetitcolas) [inline lib/drawer/drops.js]: I'll do it in following PRs, unit testing each drawer indepedently.

**Comment by jpetitcolas** [comment]
Code review either applied or delayed in another PR.
